include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -1254,38 +1664,375 @@ namespace internal
                 else
                   out[stride*n_cols] += r0;
               }
+            if (one_line == false)
+              {
+                in  += 1;
+                out += 1;
+              }
+          }
+        if (one_line == false)
+          {
+            in += stride * (mm-1);
+            out += stride * (nn-1);
+          }
+      }
+  }
 
-            // increment: in regular case, just go to the next point in
-            // x-direction. If we are at the end of one chunk in x-dir, need to
-            // jump over to the next layer in z-direction
-            switch (direction)
+
+
+  /**
+   * Internal evaluator for 1d-3d shape function using the tensor product form
+   * of the basis functions.
+   *
+   * This class implements an approach similar to the even-odd decomposition
+   * but with a different type of symmetry. In this case, we assume that a
+   * single shape function already shows the symmetry over the quadrature
+   * points, rather than the complete basis that is considered in the even-odd
+   * case. In particular, we assume that the shape functions are ordered as in
+   * the Legendre basis, with symmetric shape functions in the even slots
+   * (rows of the values array) and point-symmetric in the odd slots. Like the
+   * even-odd decomposition, the number of operations are N^2/2 rather than
+   * N^2 FMAs (fused multiply-add), where N is the number of 1D dofs. The
+   * difference is in the way the input and output quantities are symmetrized.
+   *
+   * @tparam dim Space dimension in which this class is applied
+   * @tparam n_rows Number rows in the transformation matrix, which corresponds
+   *                to the number of 1d shape functions in the usual tensor
+   *                contraction setting
+   * @tparam n_columns Number of columns in the transformation matrix, which
+   *                   corresponds to the number of 1d shape functions in the
+   *                   usual tensor contraction setting
+   * @tparam Number Abstract number type for input and output arrays
+   * @tparam Number2 Abstract number type for coefficient arrays (defaults to
+   *                 same type as the input/output arrays); must implement
+   *                 operator* with Number and produce Number as an output to
+   *                 be a valid type
+   */
+  template <int dim, int n_rows, int n_columns, typename Number, typename Number2>
+  struct EvaluatorTensorProduct<evaluate_symmetric_hierarchical,dim,n_rows,n_columns,Number,Number2>
+  {
+    static const unsigned int dofs_per_cell = Utilities::fixed_int_power<n_rows,dim>::value;
+    static const unsigned int n_q_points = Utilities::fixed_int_power<n_columns,dim>::value;
+
+    /**
+     * Empty constructor. Does nothing. Be careful when using 'values' and
+     * related methods because they need to be filled with the other
+     * constructor passing in at least an array for the values.
+     */
```

Comment:
Does this mean that the object constructed is never usable?

Reply:
You can only use the (static) function `apply`.